src/main/java/com/rtb/projectmanagementtool/task/TaskController.java

```diff
@@ -42,6 +43,77 @@ private void addKeysToTasks(ArrayList<TaskData> tasks, ArrayList<Key> keys) {
     }
   }
 
+  // Update methods
+
+  public void addUser(long taskID, long userID) {
+    addUser(getTaskByID(taskID), userID);
+  }
+
+  public void addUser(TaskData task, long userID) {
+    if (task.getTaskID() != 0) {
+      TransactionOptions options = TransactionOptions.Builder.withXG(true);
+      Transaction transaction = datastore.beginTransaction(options);
+      try {
+        addUserRecursive(task, userID);
+        transaction.commit();
+      } finally {
+        if (transaction.isActive()) {
+          transaction.rollback();
+        }
+      }
+    }
+  }
+
+  private void addUserRecursive(TaskData task, long userID) {
+    if (task.getTaskID() != 0) {
+      if (!task.getUsers().contains(userID)) {
+        task.getUsers().add(userID);
+        datastore.put(task.toEntity());
+        if (task.getParentTaskID() != 0) {
+          addUser(task.getParentTaskID(), userID);
```

Comment:
This would add a user on a subtask to all its ancestors.  This is an interesting concept.  Can you describe some use cases?

I'm thinking for example if I have a task called "Facilitate Capstone project", then I create a subtask called "Build Capstone features" which I assign to you, IMO you shouldn't need to be on my parent "Facilitate" task.

Reply:
I guess that depends on the definition of being a user assigned to a task. I figured that it's more like whether a user is working on that task (since subtask generally implies being a feature of the parent task, working on a subtask would mean working on the parent task). It seems like your definition is whether a user is responsible for completing that task.

I can change is so that the `addUser()` and `removeUser()` methods are not recursive, but I'd like to hear more opinions first so that I'm not undoing and redoing everything.

Reply:
Good points.  Probably either document on the design doc and ask for comments, or post on chat.

One other thing to consider, who's allowed to mark a status as Completed?  If we have subtask owners listed on the parent task as a user, will we need the concept of task-owners or task-admins?